crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,305 @@
+                respond_to_producer,
+            }) => {
+                self.partition_pool
+                    .buffer(record, respond_to_producer)
+                    .await?;
+            }
+            Event::Command(DispatcherMsg::Flush(respond_to_producer)) => {
+                let span = debug_span!("flush", reason = "user_flush");
+                self.partition_pool.flush_all().instrument(span).await?;
+
+                // Dropping the sender causes rx.recv() to yield with Err.
+                // This itself is used as the message, and ensures exactly one notification
+                drop(respond_to_producer);
+            }
+            Event::Timeout => {
+                let span = debug_span!("flush", reason = "timeout");
+                self.partition_pool.flush_all().instrument(span).await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub(crate) struct PartitionPool {
+    /// Pool for communicating with SPUs.
+    spus: Arc<SpuPool>,
+    /// Config for producer/flushing behavior
+    config: Arc<ProducerConfig>,
+    /// A channel for sending the status of batches
+    broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// Pool of individual `PartitionProducer`s, which oct independently
+    pool: HashMap<ReplicaKey, PartitionProducer>,
+}
+
+impl PartitionPool {
+    /// Create a new `PartitionPool`
+    pub fn new(
+        spus: Arc<SpuPool>,
+        config: Arc<ProducerConfig>,
+        broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    ) -> Self {
+        Self {
+            spus,
+            config,
+            pool: Default::default(),
+            broadcast_status,
+        }
+    }
+
+    /// Buffer the given record by sending it to the appropriate `PartitionProducer`.
+    ///
+    /// If a `PartitionProducer` does not exist for this record's `ReplicaKey`, one
+    /// will be created.
+    pub async fn buffer(
+        &mut self,
+        record: AssociatedRecord,
+        respond_to_caller: Sender<ProducerMsg>,
+    ) -> Result<(), FluvioError> {
+        let partition_producer = match self.pool.get_mut(&record.replica_key) {
+            Some(partition_producer) => partition_producer,
+            None => {
+                let partition_producer = PartitionProducer::new(
+                    self.spus.clone(),
+                    self.config.clone(),
+                    record.replica_key.clone(),
+                    self.broadcast_status.clone(),
+                );
+                self.pool
+                    .insert(record.replica_key.clone(), partition_producer);
+                self.pool
+                    .get_mut(&record.replica_key)
+                    .expect("just inserted, element must exist")
+            }
+        };
+
+        partition_producer.buffer(record, respond_to_caller).await?;
+        Ok(())
+    }
+
+    /// Flush the buffers for all partitions.
+    pub async fn flush_all(&mut self) -> Result<(), FluvioError> {
```

Comment:
this would flush all replicas which is not something desired.  should flush only specific topic

Reply:
Currently, TopicProducer and Dispatcher are 1:1, so flush_all is just flushing all the partitions for that one topic